### PR TITLE
fix(ui): only allow hugepages and core pinning for LXD pods

### LINKED ISSUE #2456: Hugepages/pinning available for virsh and lack validation
Bug originally filed by bjornt at https://bugs.launchpad.net/bugs/1922107

This is with 3.0.0b1.

If I go to the compose page for a virsh VM host, I get the option of using hugepages, as well as pinning cores.

However, we don't support that for virsh, so it shouldn't show up there.

Further more, there were no hugepages available in the system, so checking "hugepages" shouldn't have been possible either. We need to check that there is enough free hugepages for the VM to be composed.

### PR DESCRIPTION
**Pull request by Caleb-Ellis**
## Done

- only allow hugepages and core pinning for LXD pods
- disable hugepage checkbox if none are free

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Go to the KVM details page of a virsh KVM
- Open the Compose form and check that the hugepages checkbox and pin cores radio button are disabled with tooltips
- Go to the KVM details page of a LXD KVM without any free hugepages (e.g. bolla)
- Check that the hugepages checkbox is disabled

## Fixes

Fixes #2456

### CODE REVIEW

File: ui/src/app/kvm/components/KVMActionFormWrapper/ComposeForm/ComposeFormFields/ComposeFormFields.test.tsx [hunk 1]
```diff
@@ -165,6 +165,80 @@ describe("ComposeFormFields", () => {
...
+    const state = { ...initialState };
+    const store = mockStore(state);
+    const wrapper = mount(
+      <Provider store={store}>
+        <Formik initialValues={{}} onSubmit={jest.fn()}>
+          <ComposeFormFields
+            architectures={[]}
+            available={{
+              cores: 2,
+              hugepages: 0,
```
Comment (anthonydillon) [inline ui/src/app/kvm/components/KVMActionFormWrapper/ComposeForm/ComposeFormFields/ComposeFormFields.test.tsx:178]: Do you need to add some free huge pages here as you could just be testing the same mechanic twice?

File: ui/src/app/kvm/components/KVMActionFormWrapper/ComposeForm/ComposeFormFields/ComposeFormFields.tsx [hunk 2]
```diff
@@ -77,11 +93,17 @@ export const ComposeFormFields = ({
...
           wrapperClassName="u-sv2"
         />
-        <FormikField
-          label="Enable hugepages"
-          name="hugepagesBacked"
-          type="checkbox"
-        />
+        <Tooltip
+          data-test="hugepages-tooltip"
+          message={getHugepagesTooltip(isLxd, hasFreeHugepages)}
```
Comment (anthonydillon) [inline ui/src/app/kvm/components/KVMActionFormWrapper/ComposeForm/ComposeFormFields/ComposeFormFields.tsx:98]: Does an empty tooltip should a blank black box?

File: ui/src/app/kvm/components/KVMActionFormWrapper/ComposeForm/ComposeFormFields/ComposeFormFields.tsx [hunk 3]
```diff
@@ -10,27 +10,43 @@ import FormikField from "app/base/components/FormikField";
...
     memory: number; // MiB
     pinnedCores: number[];
   };
   defaults: ComposeFormDefaults;
+  podType: PodType;
+};
+
+const getHugepagesTooltip = (isLxd: boolean, hasFreeHugepages: boolean) => {
+  if (!isLxd) {
+    return "Hugepages are only supported on LXD KVMs.";
```
Comment (anthonydillon) [inline ui/src/app/kvm/components/KVMActionFormWrapper/ComposeForm/ComposeFormFields/ComposeFormFields.tsx:30]: We often mix the use of KVM and VM. In this case does LXD refer to them as KVMs?

File: ui/src/app/kvm/components/KVMActionFormWrapper/ComposeForm/ComposeFormFields/ComposeFormFields.test.tsx (hunk 1 above)
Comment (Caleb-Ellis) [reply ui/src/app/kvm/components/KVMActionFormWrapper/ComposeForm/ComposeFormFields/ComposeFormFields.test.tsx:178]: I don't think so - this test is just for non-LXD pods so adding any here won't change anything. Plus I think it's not possible to have this number be anything other than 0 for non-LXD pods?

File: ui/src/app/kvm/components/KVMActionFormWrapper/ComposeForm/ComposeFormFields/ComposeFormFields.tsx (hunk 3 above)
Comment (Caleb-Ellis) [reply ui/src/app/kvm/components/KVMActionFormWrapper/ComposeForm/ComposeFormFields/ComposeFormFields.tsx:30]: Yeah this refers to them as KVMs.

File: ui/src/app/kvm/components/KVMActionFormWrapper/ComposeForm/ComposeFormFields/ComposeFormFields.tsx (hunk 2 above)
Comment (Caleb-Ellis) [reply ui/src/app/kvm/components/KVMActionFormWrapper/ComposeForm/ComposeFormFields/ComposeFormFields.tsx:98]: Nah if the message is falsy it doesn't render anything

**Review by huwshimi — APPROVED** [review]

**Pull request merged**